src/d3a/blockchain/contracts/ink/lib.rs

```diff
@@ -0,0 +1,203 @@
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use ink_lang as ink;
+
+#[ink::contract(version = "0.1.0")]
+mod erc20 {
+    use ink_core::storage;
+
+    #[ink(storage)]
+    struct Erc20 {
+        /// The total supply.
+        total_supply: storage::Value<Balance>,
+        /// The balance of each user.
+        balances: storage::HashMap<AccountId, Balance>,
+        /// Approval spender on behalf of the message's sender.
+        allowances: storage::HashMap<(AccountId, AccountId), Balance>,
+        /// The balance of each user.
+        trades: storage::HashMap<Hash, (AccountId, AccountId, Balance, i8)>,
```

Comment:
Can this not be of type `storage::HashMap<Hash, Trade>,` ? Is the tuple needed here?

Reply:
Depends on which data we want to see in the response from the contract call! What type would the Trade be then?

Reply:
If I understand the contract correctly though, this is the type of the storage, therefore it is the type that the Trade objects are going to be stored. The response from the contract call is defined by the get_trades contract function, which has a return value of a tuple indeed, therefore theoretically we could store the trade as a Trade object, and convert it to a tuple on the get_trades method. That way inside the smart contract, we will be aware that Trades are dealt with using their own type.

In addition to that, I am intrigued to see if Ink provides a standardized way to handle structs as contract return values. Not really important though, we can live without this functionality